### PR TITLE
Fixes #85

### LINKED ISSUE #85: Casting a lambda expression causes a parsing failure
I failed to parse some Java 8 files with lambda expressions; my stack trace was something like this

```
java.lang.NullPointerException
    at com.github.javaparser.ast.Node.setAsParentNodeOf(Node.java:331)
    at com.github.javaparser.ast.expr.MethodCallExpr.setArgs(MethodCallExpr.java:96)
    at com.github.javaparser.ast.expr.MethodCallExpr.<init>(MethodCallExpr.java:63)
    at com.github.javaparser.ASTParser.PrimarySuffixWithoutSuper(ASTParser.java:3180)
```

Obviously, the provided arguments was a list containing a `null` value. I boiled it down to a simple statement that cannot be parsed:

```
Stream<CharSequence> stream = Stream.generate((Supplier<CharSequence>) () -> "Fail");
```

Removing the cast makes the problem go away.

As this is certainly to be fixed within the `java_1_8.jj` file, I won't come up with a fix (this stuff is just beyond me). I can however set up a branch with a test case or something.

### PR DESCRIPTION
**Pull request by sebastiankirsch**

### CODE REVIEW

File: javaparser-core/src/main/javacc/java_1_7.old [hunk 1]
```diff
@@ -1,20 +1,19 @@
 /*
```
Comment (ftomassetti) [inline javaparser-core/src/main/javacc/java_1_7.old:1]: Why do we have a file named "javaparser-core/src/main/javacc/java_1_7.old" ?

Comment (sebastiankirsch) [inline javaparser-core/src/main/javacc/java_1_7.old:1]: @SmiddyPence kept it when implementing #27. It was quite useful to hunt down & understand the lambda stuff tbh. I'd keep it until we add Java 9 support :)

**Comment by sebastiankirsch** [comment]
Although the test case is fine, real life still has issues with casted lambdas. Will try to come up with a more elaborate (failing) test case.

**Comment by rpau** [comment]
Ok, leave me the tests and I can work on it :)

**Comment by sebastiankirsch** [comment]
Can I just delete the comment and we forget about that? Was just me and SNAPSHOT versions and local repo and pom files and stuff :/

Everything working just fine!

File: javaparser-core/src/main/javacc/java_1_7.old [hunk 2]
```diff
@@ -1285,6 +1256,8 @@ List ExtendsList(boolean isInterface):
    boolean extendsMoreThanOne = false;
    List ret = new LinkedList();
    ClassOrInterfaceType cit;
+    AnnotationExpr ann;
```
Comment (ftomassetti) [inline javaparser-core/src/main/javacc/java_1_7.old:1259]: has line 1259 an extra space?

File: javaparser-core/src/main/javacc/java_1_7.old [hunk 3]
```diff
@@ -1824,16 +1729,13 @@ WildcardType Wildcard():
 		"super" sup = ReferenceType()
    ]
    { 
-       WildcardType tmp = new WildcardType(line, column, token.endLine, token.endColumn,ext, sup); 
-        
-       return tmp;
+     return new WildcardType(line, column, token.endLine, token.endColumn,ext, sup);
```
Comment (ftomassetti) [inline javaparser-core/src/main/javacc/java_1_7.old:1732]: Collapse that and the brackets in one line

File: javaparser-core/src/main/javacc/java_1_7.old [hunk 4]
```diff
@@ -1937,18 +1824,15 @@ Expression Expression():
 	Expression ret;
 	AssignExpr.Operator op;
 	Expression value;
-       
 }
 {
   ret = ConditionalExpression()
   [
-    LOOKAHEAD(2)
+       LOOKAHEAD(2)
```
Comment (ftomassetti) [inline javaparser-core/src/main/javacc/java_1_7.old:1831]: I think it is better without the extra space

File: javaparser-core/src/main/javacc/java_1_7.old [hunk 5]
```diff
@@ -2297,53 +2113,41 @@ Expression CastExpression():
 	Type type;
 	int line;
 	int column;
-       
 }
 {
-  "(" {line=token.beginLine; column=token.beginColumn;} 
+  "(" {line=token.beginLine; column=token.beginColumn;}
```
Comment (ftomassetti) [inline javaparser-core/src/main/javacc/java_1_7.old:2118]: maybe add spaces around the brackets

File: javaparser-core/src/main/javacc/java_1_7.old [hunk 6]
```diff
@@ -2360,7 +2164,7 @@ Expression PrimaryPrefix():
 	  "this" { ret = new ThisExpr(token.beginLine, token.beginColumn, token.endLine, token.endColumn, null); }
 	|
 	  "super" { ret = new SuperExpr(token.beginLine, token.beginColumn, token.endLine, token.endColumn, null); }
-	  "."
+	     "."
```
Comment (ftomassetti) [inline javaparser-core/src/main/javacc/java_1_7.old:2167]: do we want the extra space?

File: javaparser-core/src/main/javacc/java_1_7.old [hunk 7]
```diff
@@ -2375,13 +2179,13 @@ Expression PrimaryPrefix():
 				ret = f;
 			}
 	  	}
-	|
+	 |
```
Comment (ftomassetti) [inline javaparser-core/src/main/javacc/java_1_7.old:2182]: extra space

File: javaparser-core/src/main/javacc/java_1_7.old [hunk 8]
```diff
@@ -2375,13 +2179,13 @@ Expression PrimaryPrefix():
...
 	  "(" {line=token.beginLine; column=token.beginColumn;} ret = Expression() ")" { ret = new EnclosedExpr(line, column, token.endLine, token.endColumn,ret); }
 	|
 	  ret = AllocationExpression(null)
 	|
 	  LOOKAHEAD( ResultType() "." "class" )
-	  type = ResultType() "." "class" { ret = new ClassExpr(type.getBeginLine(), type.getBeginColumn(), token.endLine, token.endColumn, type); }
+	  type = ResultType()  "." "class" { ret = new ClassExpr(type.getBeginLine(), type.getBeginColumn(), token.endLine, token.endColumn, type); }
```
Comment (ftomassetti) [inline javaparser-core/src/main/javacc/java_1_7.old:2188]: extra space

File: javaparser-core/src/main/javacc/java_1_7.old [hunk 9]
```diff
@@ -2873,9 +2600,8 @@ IfStmt IfStatement():
 	Statement elseStmt = null;
 	int line;
 	int column;
-       
-       Comment thenCmmt = null;
-       Comment elseCmmt = null;
+    Comment thenCmmt = null;
```
Comment (ftomassetti) [inline javaparser-core/src/main/javacc/java_1_7.old:2603]: it should be aligned with previous lines

File: javaparser-core/src/main/javacc/java_1_7.old [hunk 10]
```diff
@@ -2873,9 +2600,8 @@ IfStmt IfStatement():
...
+    Comment elseCmmt = null;
```
Comment (ftomassetti) [inline javaparser-core/src/main/javacc/java_1_7.old:2604]: Not sure why we want to refer to comments in the grammar

File: javaparser-core/src/main/javacc/java_1_8.jj [hunk 11]
```diff
@@ -2785,40 +2782,18 @@ ExpressionStmt StatementExpression():
...
-	       params = add(params, new Parameter(expr.getBeginLine(), expr.getBeginColumn(), expr.getEndLine(), expr.getEndColumn(), 0, null, null, false, id));
-	       expr = new LambdaExpr(expr.getBeginLine(), expr.getBeginColumn(), expr.getEndLine(), expr.getEndColumn(), params, lambdaBody, false);
-	     }
-	     else if( expr instanceof LambdaExpr)
+	     if (expr instanceof CastExpr)
 	     {
-	       ((LambdaExpr)expr).setBody(lambdaBody);
-	       expr.setEndLine(lambdaBody.getEndLine());
-	       expr.setEndColumn(lambdaBody.getEndColumn());
+           inner = generateLambda(((CastExpr)expr).getExpr(), lambdaBody);
```
Comment (ftomassetti) [inline javaparser-core/src/main/javacc/java_1_8.jj:2787]: should be the content of the brackets one tab inside?

**Comment by ftomassetti** [comment]
+1

**Comment by sebastiankirsch** [comment]
The changes in java_1_7.old were made simply to be able to diff it with the java_1_8.jj file and "see" only the changes to the grammar. My intention was to keep the changes to java_1_8.jj as minimal as possible in this pull request. I will happily apply some "modern" formatting of the java_1_8.jj file once this pull request is merged :)

**Comment by ptitjes** [comment]
This will be painfull to maintain the 1.7 grammar without even testing it... I say grop it! :P
This is why we have to simplify the grammar and limit the java code inside it to bare minimal.

**Comment by sebastiankirsch** [comment]
The 1.7 grammar is not used any more. It's just there (and I found it quite useful to understand the changes from 7 -> 8).
I totally agree with your approach of taking as much out of the .jj file as possible!

**Comment by ptitjes** [comment]
@sebastiankirsch As for the method for lambda, here is what I thought during the night:
- we could either make it a static method of LambdaExpr as it contains no reference to parser
- or we can leave in there and make it go in ASTParserBase when my pull request has been merged
  What do you think ?

As for the 1.7 grammar, maybe we could let javacc generate the actual parser (in a separate dir) as to be able to apply the same evolutions/simplifications as in the 1.8 one. In a way, it is just a subset of the trees the 1.8 parser builds. So we could also tag the individual tests (for java version) to run them against the 1.7 parser too. The question is: is it worth it ?
My personal goal as for (*) javaparser is to gain enough confidence and knowledge about the way the grammar works to be able to simplify and document it so that fixing future bugs like the present one could be fixed by anybody.

(*) That is to say as for the parser part of javaparser if we all agree to include in the project the model part that we started to develop with @ftomassetti.

**Comment by ptitjes** [comment]
@sebastiankirsch @ftomassetti @matozoid I see no show stopper for this. I vote (if I may say my word :p) for it to be merged.

**Comment by sebastiankirsch** [comment]
@ptitjes concerning the lambda stuff, I'd like refer to [PMD](http://sourceforge.net/p/pmd/code/ci/master/tree/pmd-java/etc/grammar/Java.jjt) again: to my untrained eye, not only the lambda solution looks much smoother. Also the fact that they handle lambdas only once and not twice as we do is another indication that someone other there knows what has to be done.
I think it would be a good idea to "copy" as much of their grammar, assuming that they will support Java 9 quickly, giving us much easier migration paths than we experienced with Java 8.

**Comment by ptitjes** [comment]
@sebastiankirsch I'm willing to get into their grammar and get our grammar closer to it if you want.

**Pull request merged**